### ISSUE TITLE
Console problem - "Plugin BakingPlate could not be found."

### ISSUE THREAD

**Issue by sdd** [ProLoser/BakingPlate#16]
Hi Guys, I'm having a bit of an issue integrating BakingPlate with my new workflow. I want to have CakePHP itself as a submodule within my project for ease of updates. I'm using CakePHP 2.0. So, I created a blank directory, ran `git init` from inside it, then added cakephp as a submodule in a folder called cakephp in the root of my project. I then copied app/, index.php, plugins/, .htaccess, and vendors/ from cakephp to the project root so that I can leave the cakephp folder alone. I edited /index.php and app/webroot/index.php so that CAKE_CORE_INCLUDE_PATH points to `ROOT . DS . 'cakephp' . DS . 'lib'`. 

So far, so good - with apache's DocumentRoot pointed at the project root, cake's default home displays ok.

Now my problem comes when I try to integrate BakingPlate with this. I want to add BP as a git submodule as well - again for ease of upgrade, deployment, etc. I tried adding BP as a submodule using `git submodule add git://github.com/sams/BakingPlate.git plugins/baking_plate`. By cd'ing into /app and executing `../cakephp/lib/Cake/Console/cake`, the cake shell worked, but plate did not show as an available shell. I tried adding symbolic links in /app/Console/Command to /plugins/baking_plate/Console/Command/PlateShell.php and in /app/Console/Templates to the folders in /plugins/baking_plate/Console/Templates. This resulted in plate showing up in the list of available shells from the cake console, and cake plate now shows the cake plate subcommands correctly. But if I try cake plate bake or cake plate browse, I get the following errors:

```
Error: Plugin BakingPlate could not be found.
#0 /Users/Liquid/workspace/cashback_us/cakephp/lib/Cake/Core/App.php(332): CakePlugin::path('BakingPlate')
#1 /Users/Liquid/workspace/cashback_us/cakephp/lib/Cake/Configure/PhpReader.php(71): App::pluginPath('BakingPlate')
#2 /Users/Liquid/workspace/cashback_us/cakephp/lib/Cake/Core/Configure.php(309): PhpReader->read('BakingPlate.sub...')
#3 /Users/Liquid/workspace/cashback_us/app/Plugin/BakingPlate/Console/Command/PlateShell.php(101): Configure::load('BakingPlate.sub...')
#4 /Users/Liquid/workspace/cashback_us/cakephp/lib/Cake/Console/Shell.php(199): PlateShell->_welcome()
#5 /Users/Liquid/workspace/cashback_us/cakephp/lib/Cake/Console/Shell.php(370): Shell->startup()
#6 /Users/Liquid/workspace/cashback_us/cakephp/lib/Cake/Console/ShellDispatcher.php(177): Shell->runCommand('browse', Array)
#7 /Users/Liquid/workspace/cashback_us/cakephp/lib/Cake/Console/ShellDispatcher.php(69): ShellDispatcher->dispatch()
#8 /Users/Liquid/workspace/cashback_us/cakephp/lib/Cake/Console/cake.php(24): ShellDispatcher::run(Array)
#9 {main}
```

I have tried having baking plate as a submodule in /plugins/BakingPlate, /app/Plugin/BakingPlate, and /app/Plugin/baking_plate, but all result in the same error.

Thanks, Scotty

**Comment by ProLoser** [comment]
Just an FYI, Sams fork and mine vary greatly.

Anywho, since you symlinked only a subfolder, that is why you're getting most of the problems. I'm having a little difficulty visualizing your folder structure, but I have something like this:
- some_app/[1 app folder as a git repo]
- another_app/[2 app folder as another git repo]
- plugins/[BakingPlate]
- vendors
- cakephp/[cake as a git repo]

Or some variant of that ^ (needs tweaking I think). Anywho, don't bother rearranging your folders, what you should try doing is symlinking your cakephp/plugins > /pugins
You shouldn't be simply symlinking subfolders of the plugin.

You should also add the path to the cake console to your $PATH (search tv.cakephp.com for adding cake to your console). I too have run into complexities with using the plate plugin outside the app directory (for baking new projects) but you can pass the path to the plugins folder as an argument. 

If that's not enough info I can try some stuff myself

**Comment by sams** [comment]
I had to load BakingPlate within the bootstrap of the app within my cake 2.0 install (as well as cake on paths)

I have one cake 1.3 and one 2.x install

then when baking a project I set the full path when prompted - otherwise I get projects created within either the cake install or the app directory of the cake install

alias added to profile or whatever
`alias plate="cake  BakingPlate.plate  -app /path/to/cake/2.0/app"`

in `/path/to/cake/2.0/app/Config/bootstrap.php`
`CakePlugin::load('BakingPlate');`

this means I can now `plate add users -g cakedc` or `plate search` etc instead of `cake BakingPlate.plate ...`

Use Proloser BakingPlate - I am expecting a major push over the weekend

**Comment by ProLoser** [comment]
I was wondering why all your forks were gone.

On Thu, Dec 15, 2011 at 3:59 PM, Sam Sherlock <
reply@reply.github.com

**Comment by sams** [comment]
I felt removing stuff and much more selective about what to put in the skel was the way forward
I removed loads of submodules too for the same reason but will be adding some back in (also I felt that if they are on cakeplugins.com then submodules could be really reduced)

I found that aliasing helpers has some quirks (thats why I had all that chaos in my 2.0 branch recently with the helpers)
http://www.dereuromark.de/2011/10/31/freshly-baked-cake2-0-tips/#comments

**Comment by ProLoser** [comment]
You do realize his code example is broken right? The file/class name is
HtmlPublicHelper and you guys are all using PublicHtml, which is backwards.
My version has worked perfectly fine and I have not encountered any issues.

On Thu, Dec 15, 2011 at 4:48 PM, Sam Sherlock <
reply@reply.github.com

**Comment by sams** [comment]
haha good spot that. Still though what I am saying applies. (I guess he had a typo)

Your setup works even when Html or Form are redeclared in SomeOtherController?
I found that alter functionality was only available when the alias set in the AppController and not in SomeOtherController (or maybe aliased also but thats excessive)

In somecases the extra functionality eg HtmlPlusHelper::time was not available in the View Template - whereas it was available in the Layout.  $this->Html was HtmlHelper in View and HtmlPlusHelper in Layout

Another thing to notice is that if a plugin sets a css file inline false then the link tag will be added to the scripts_for_layout var and then output before  body not in the head.

Have you found this or is it just me? (hoping you prove me wrong :))

**Comment by sdd** [comment]
Hi guys, thanks for the input - much appreciated. I'll go through your suggestions and see if I can make any progress.

**Comment by sdd** [comment]
Hi Guys,

Looks like the issue was as straightforward as loading the plugin in bootstrap.php, thanks. I changed the symlinking as recommended also. 

The only remaining issue was that plate's git clone operation failed due to the symlinks. I think I will have a bit of a play around to see if I can get a better arrangement of git,submodules,symlinks, cake location, etc

**Comment by ProLoser** [comment]
Personally, I don't use any symlinks. I believe there are ways to make it
behave properly without symlinking at all. Also, yes, shoulda mentioned
that you now gotta load plugins explicitly in the bootstrap. That was a
pain for me too.

The reason I don't have a defacto answer is because I am trying to decide
on the best way to make the plugin globally accessible (aka: when you're
not inside a project) for baking new projects. Need to do a little more
research. Until then, the .profile alias Sam suggested should suffice as a
decent shortcut.

Note: Scott, checkout my fork, it's a lot farther along than sams fork
(currently) and has some awesome new features like CakePackages.com
integration.

On Fri, Dec 16, 2011 at 7:51 AM, Scott Donnelly <
reply@reply.github.com

**Comment by sdd** [comment]
Hi ProLoser, I was actually using your fork, It says sams in my question because I copy and pasted it from the wiki :-) 

I was looking into whether to have a global copy of plugins and of cake as well. I've decided not to - I seem to be frequently changing the machine I am using for developing, so I am instead going down the route of having a skeleton repo that includes both cake itself and baking plate as submodules. I am less concerned with having multiple copies of things on the same machine, so I don't need to worry so much about global accessibility. I want to be able to clone my skeleton repo, execute git submodule update --init --recursive, and have it pull the latest stable cake 2.0, baking plate, and maybe a few other things - and them be working immediately with no need to change configs / move files etc. I've got it mostly working now, as well as a bash script that creates a vhost in apache and adds a entry in hosts. The only tricky bit is the cake console - getting this to "just work" after a clone onto a new machine will need another script. my skeleton at the moment has baking plate configured and all of the modules installed (but not all added in bootstrap.php) - this means I can start a new project on a new machine and have them all there without worrying about getting the plate console working on that machine. Phew, this message turned out much longer than I thought it would be!
